Сделано 2 задание

Krayseer opened the pull request.

Avrooid committed:
Добавил поля gender, age, block в Contact, для фильтрации поиска и сортировки, реализовал SortService, исправил ошибки в ContactService и GroupService

Krayseer committed:
Добавлены сервисы по поиску и фильтрации контактов, обновлены тестовые классы и список команд

Avrooid committed:
Добавлены сервисы поиска, сортировки и фильтрации контактов. В контакты добавлены поля возраст, пол, блокировка

vpyzhyanov commented on src/main/java/ru/anykeyers/domain/Contact.java:
```diff
@@ -28,6 +30,20 @@ public class Contact {
+    /**
+     * Блокировка контакта
+     */
+    private boolean block;
```
хорошо бы сделать отступ перед конструктором, чтобы поле с ним не сливалось

Krayseer replied:
сделано

vpyzhyanov commented on src/main/java/ru/anykeyers/domain/Contact.java:
```diff
@@ -28,6 +30,20 @@ public class Contact {
+    /**
+     * Блокировка контакта
+     */
+    private boolean block;
```
boolean свойства принято называть с помощью вопросительных глаголов, "is", "has" или простых страдательных глаголов типа "enabled". В данном случае можно назвать поле "bloked" или "isBlocked"

Krayseer replied:
исправлено

vpyzhyanov commented on src/main/java/ru/anykeyers/domain/Contact.java:
```diff
@@ -70,18 +86,59 @@ public void setPhoneNumber(String phoneNumber) {
+     * @return человекочитаемая информация о контакте
+     */
+    public String getInfo() {
+        StringBuilder builder = new StringBuilder();
```
Разве вы не видите предупреждение в IDEA? Она говорит вам, что тут можно использовать просто String. И она права.

Krayseer replied:
убрано в UiContactMapper

vpyzhyanov commented on src/main/java/ru/anykeyers/domain/Contact.java:
```diff
@@ -70,18 +86,59 @@ public void setPhoneNumber(String phoneNumber) {
+    /**
+     * @return человекочитаемая информация о контакте
+     */
+    public String getInfo() {
```
По хорошему форматированием в читаемое строковое представление должен заниматься отдельный класс. Так можно делать только, если вы хотите сделать настоящую инкапсуляцию, но вы даже не пытались. Так что однозначно в отдельный класс.

Тут я немного придираюсь, т.к. такая задача может и не стоит вынесения в отдельный класс, но это важно в учебных целях, чтобы вы задумывались о том, что разные задачи нужно разделять по разным классам. Так что надо исправить.

Krayseer replied:
Убрано в UiContactMapper

vpyzhyanov commented on src/main/java/ru/anykeyers/domain/Contact.java:
```diff
@@ -70,18 +86,59 @@ public void setPhoneNumber(String phoneNumber) {
+        StringBuilder builder = new StringBuilder();
+        return builder
+                .append("Имя: ").append(name)
+                .append("\nНомер телефона: ").append(phoneNumber != null ? phoneNumber : StringUtils.EMPTY)
```
ставить символ перехода на следующую строку **в начале строки** - дурной тон. Он должен быть в конце строки. Рекомендую вам использовать фичу java17 для вывода многострочного текста с подстановкой параметров.

Krayseer replied:
исправили, спасибо

vpyzhyanov commented on src/main/java/ru/anykeyers/domain/Contact.java:
```diff
@@ -70,18 +86,59 @@ public void setPhoneNumber(String phoneNumber) {
+        return age == contact.getAge() &&
+                block == contact.isBlock() &&
+                Objects.equals(id, contact.getId()) &&
+                Objects.equals(username, contact.getUsername()) &&
```
метод реализован по прежнему некорректно. Это замечание было в первой задаче! Пожалуйста, больше не допускайте неисправленных замечаний из предыдущей задачи.

Krayseer replied:
теперь там только id контакта

vpyzhyanov commented on src/main/java/ru/anykeyers/domain/Group.java:
```diff
@@ -91,6 +91,13 @@ public boolean deleteContactFromGroup(Contact contact) {
+     * Получить информацию по группе
+     */
+    public String getInfo() {
+        return String.format("Название: %s. Количество участников: %s.", name, contacts == null ? 0 : contacts.size());
```
длинная строчка, перенесите все параметры на следующую строку. Особенно с учётом того, что в одном параметре условие.

Не увлекайтесь однострочным программированием, строчки - бесплатные)

**ВАЖНО! В вашем случае допускать null является большой ошибкой**, особенно если учесть то, что вам ничего не стоит гарантировать, что тут будет не null. Предлагаю вам полностью избавиться от null. null - это потенциальный неожиданный NullPointerException

Krayseer replied:
исправили недочет

vpyzhyanov commented on src/main/java/ru/anykeyers/domain/User.java:
```diff
@@ -34,14 +34,9 @@ public class User {
-     * Название группы, выбранное для редактирования
-     */
-    private String groupNameToEdit;
+    private String editInfo;
```
Насколько я понял это поле - является параметром, используемым при обработке состояния. Добро пожаловать на тёмную сторону ООП, ваш класс `User` решает несколько задач. Явно тут нужен ещё один класс. Как думаете, какой?

vpyzhyanov replied:
Нет ответа!!!

Krayseer replied:
Добавил сервис UserStateService для работы с состояниями пользователя

vpyzhyanov commented on src/main/java/ru/anykeyers/domain/User.java:
```diff
@@ -128,13 +114,12 @@ public boolean equals(Object o) {
                 && stateType == user.getStateType()
-                && Objects.equals(contactNameToEdit, user.getContactNameToEdit())
-                && Objects.equals(groupNameToEdit, user.getGroupNameToEdit())
+                && Objects.equals(editInfo, user.getEditInfo())
```
метод equals реализован некорректно! Вы правда считаете, что пользователь, у которого изменилось состояние - это уже другой пользователь?!

Krayseer replied:
теперь там только id

vpyzhyanov commented on src/main/java/ru/anykeyers/domain/kinds/GroupEditKind.java:
```diff
@@ -0,0 +1,23 @@
+/**
+ * Аргументы для изменения группы
+ */
+public final class GroupEditKind {
```
описание и название класса не соответствуют друг другу. Название говорит про какой-то вид группы, а описание про аргументы. А значения - вообще числа. Я ничего не понял

Krayseer replied:
Этот и все следующие классы по типу этого были переименованы, добавлен новый javadoc

vpyzhyanov commented on src/main/java/ru/anykeyers/domain/kinds/SearchKind.java:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Аргументы для поиска контактов
+ */
+public final class SearchKind {
```
аналогично

vpyzhyanov commented on src/main/java/ru/anykeyers/domain/kinds/contact/ContactEditBlockKind.java:
```diff
@@ -0,0 +1,18 @@
+/**
+ * Аргументы изменения блокировки контакта
+ */
+public final class ContactEditBlockKind {
```
все остальные тоже

vpyzhyanov commented on src/main/java/ru/anykeyers/factories/StateProcessorFactory.java:
```diff
@@ -48,6 +46,14 @@ public StateProcessor createStateProcessorByType(StateType stateType) {
                 GroupService groupService = new GroupServiceImpl(groupRepository, contactRepository);
                 return new GroupStateProcessor(authenticationService, groupService);
             }
+            case FUNCTION -> {
```
не очень удачное имя выбрали. Обычно под функцией понимают функцию, а не одно из состояний. Ну и это очень общий термин, сложно понять что вы тут имеете в виду. Благо есть JavaDoc, который поясняет. Но лучше отразить это пояснение в имени `(поиск, фильтр, сортировка)`, и, желательно, без слова function.

Krayseer replied:
Исправили на Operation(выполняемая операция)

vpyzhyanov commented on src/main/java/ru/anykeyers/factories/StateProcessorFactory.java:
```diff
@@ -48,6 +46,14 @@ public StateProcessor createStateProcessorByType(StateType stateType) {
+            case FUNCTION -> {
+                GroupRepository groupRepository = repositoryFactory.createGroupRepository();
+                ContactRepository contactRepository = repositoryFactory.createContactRepository();
+                SearchService searchService = new SearchServiceImpl(contactRepository, groupRepository);
```
похоже я только сейчас заметил существенную просадку по производительности и памяти. Зачем вы **каждый раз** и **на каждый запрос** создаёте сервисы и репозитории? Да ещё и обработчик состояния `ru.anykeyers.factories.StateProcessorFactory#createStateProcessorByType` оказывается тоже каждый раз создаётся заново! Не надо так. Создайте один раз и используйте. 

Например, все `StateProcessor` можно сложить в Map и доставать по типу состояния

Krayseer replied:
сделано в новой версии

vpyzhyanov commented on src/main/java/ru/anykeyers/processors/states/impl/ContactStateProcessor.java:
```diff
@@ -42,26 +43,38 @@ private void registerStatesHandlers() {
             return messages.getMessageByKey("contact.state.edit.field");
         });
         registerHandler(State.EDIT_CONTACT_FIELD, (user, field) -> {
             switch (field.toLowerCase()) {
```
странно цифру приводить к нижнему регистру)))

Krayseer replied:
и правда странно, убрано

vpyzhyanov commented on src/main/java/ru/anykeyers/processors/states/impl/ContactStateProcessor.java:
```diff
@@ -42,26 +43,38 @@ private void registerStatesHandlers() {
+                }
             }
-            return messages.getMessageByKey("argument.bad");
+            return messages.getMessageByKey("commons.bad-argument");
```
Замечание про messages не исправлено!

Krayseer replied:
в новой версии исправлено

vpyzhyanov commented on src/main/java/ru/anykeyers/processors/states/impl/FunctionalStateProcessor.java:
```diff
@@ -0,0 +1,163 @@
+        this.filterService = filterService;
+        this.sortService = sortService;
+        this.authenticationService = authenticationService;
+        messages = new Messages();
```
очередной раз читается один и тот же файл. Нехорошо

vpyzhyanov commented on src/main/java/ru/anykeyers/processors/states/impl/FunctionalStateProcessor.java:
```diff
@@ -0,0 +1,163 @@
+/**
+ * Обработчик состояний для получения обработанной информации (поиск, фильтрация, сортировка)
+ */
+public class FunctionalStateProcessor extends BaseStateProcessor {
```
аналогично: выбрано неудачное имя с использованием слова `Functional`

Krayseer replied:
-> Operation

vpyzhyanov commented on src/main/java/ru/anykeyers/processors/states/impl/FunctionalStateProcessor.java:
```diff
@@ -0,0 +1,163 @@
+     */
+    private void registerSearchStateHandlers() {
+        registerHandler(State.SEARCH_KIND, (user, message) -> {
+            switch (message.toLowerCase()) {
```
снова цифры к нижнему регистру приводятся

Krayseer replied:
исправлено

vpyzhyanov commented on src/main/java/ru/anykeyers/processors/states/impl/FunctionalStateProcessor.java:
```diff
@@ -0,0 +1,163 @@
+        messages = new Messages();
+        registerSearchStateHandlers();
+        registerFilterCommands();
+        registerSortCommands();
```
молодцы, что разбили на методы по смыслу

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/FileContactRepository.java:
```diff
@@ -47,6 +47,10 @@ public Set<Contact> findByUsername(String username) {
     @Override
     public Contact findByUsernameAndId(String username, String contactId) {
+        Set<Contact> contacts = findByUsername(username);
+        if (contacts == null) {
```
1. В JavaDoc метода `findByUsername` нет ни слова о том, что он возвращает null
2. Нет никого смысла возвращать null, когда вы используете коллекцию. Возвращайте **пустую коллекцию**. В данном случае `Set.of()`
3. Всегда старайтесь избегать null.

Krayseer replied:
исправлено

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/FileContactRepository.java:
```diff
@@ -47,6 +47,10 @@ public Set<Contact> findByUsername(String username) {
     public Contact findByUsernameAndId(String username, String contactId) {
+        Set<Contact> contacts = findByUsername(username);
+        if (contacts == null) {
+            return null;
```
в этом случае лучше использовать Optional, он создан как раз для того, чтобы методы его возвращали, для корректной и удобной обработки null значений

Krayseer replied:
так и сделали

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/FileContactRepository.java:
```diff
@@ -57,7 +61,11 @@ public Contact findByUsernameAndId(String username, String contactId) {
-        for (Contact contact : findByUsername(username)) {
+        Set<Contact> contacts = findByUsername(username);
+        if (contacts == null) {
+            return null;
```
всё аналогично

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/FileContactRepository.java:
```diff
@@ -67,7 +75,11 @@ public Contact findByUsernameAndName(String username, String name) {
 
     @Override
     public Contact findByUsernameAndPhoneNumber(String username, String phoneNumber) {
```
в таких случая принято называть метод с использованием слова `get`. Имя `find` уместно только тогда, когда вы делаете **поиск** и возвращаете коллекцию значений.

Krayseer replied:
исправили

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/FileContactRepository.java:
```diff
@@ -57,7 +61,11 @@ public Contact findByUsernameAndId(String username, String contactId) {
 
     @Override
     public Contact findByUsernameAndName(String username, String name) {
```
в таких случая принято называть метод с использованием слова `get`. Имя `find` уместно только тогда, когда вы делаете **поиск** и возвращаете коллекцию значений.

Krayseer replied:
исправили

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/FileContactRepository.java:
```diff
@@ -67,7 +75,11 @@ public Contact findByUsernameAndName(String username, String name) {
-        for (Contact contact : findByUsername(username)) {
+        Set<Contact> contacts = findByUsername(username);
+        if (contacts == null) {
+            return null;
```
аналогично

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/FileContactRepository.java:
```diff
@@ -88,7 +100,10 @@ public void saveOrUpdate(Contact contact) {
     public void delete(Contact contact) {
-        Set<Contact> contacts = contactsByUsername.get(contact.getUsername());
+        Set<Contact> contacts = findByUsername(contact.getUsername());
+        if (contacts == null) {
```
аналогично

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/FileGroupRepository.java:
```diff
@@ -52,6 +52,10 @@ public Set<Group> findByUsername(String username) {
     @Override
     public Group findByUsernameAndName(String username, String groupName) {
+        Set<Group> groups = findByUsername(username);
+        if (groups == null) {
```
аналогично

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/FileContactRepository.java:
```diff
@@ -47,6 +47,10 @@ public Set<Contact> findByUsername(String username) {
+        if (contacts == null) {
+            return null;
+        }
         for (Contact contact : findByUsername(username)) {
```
2 раза подряд вызывается одна и та же функция

Krayseer replied:
ошибка исправлена

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/FileGroupRepository.java:
```diff
@@ -52,6 +52,10 @@ public Set<Group> findByUsername(String username) {
+        if (groups == null) {
+            return null;
+        }
         for (Group group : findByUsername(username)) {
```
2 раза подряд вызывается одна и та же функция

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/parsers/FileContactParser.java:
```diff
@@ -11,9 +14,12 @@ public class FileContactParser implements FileObjectParser<Contact> {
     public String parseTo(Contact object) {
         StringBuilder builder = new StringBuilder();
```
не вполне уместное использование `StringBuilder`. Его нужно использовать только тогда, когда у вас очень много данных и вы заполняете их в цикле. Во всех остальных случаях надо испольозовать String. Оптимизатор Java сделает свою магию и обычная строка будет работать гораздо эффективнее

Krayseer replied:
Исправлено

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/parsers/FileContactParser.java:
```diff
@@ -11,9 +14,12 @@ public class FileContactParser implements FileObjectParser<Contact> {
     public String parseTo(Contact object) {
```
замечания про имена методов и класса **из предыдущей задачи** не исправлены!!!

Krayseer replied:
исправлено в новой версии

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/parsers/FileContactParser.java:
```diff
@@ -38,6 +44,16 @@ public Contact parseFrom(String line) {
+                    contact.setGender(value == null ? null : Gender.valueOf(value));
+                    break;
+                case "block":
+                    contact.setBlock(Objects.equals(value, "true"));
```
`Boolean.parseBoolean(value)`

Krayseer replied:
Поправили

vpyzhyanov commented on src/main/java/ru/anykeyers/repositories/file/parsers/FileUserParser.java:
```diff
@@ -17,8 +17,7 @@ public String parseTo(User object) {
                 .append(";bot_type=").append(object.getBotType())
-                .append(";edit_contact_name=").append(object.getContactNameToEdit())
-                .append(";edit_group_name=").append(object.getGroupNameToEdit())
+                .append(";edit_info=").append(object.getEditInfo())
```
Зачем вам выгружать состояние? Какую задачу вы этим решаете?

Теперь я знаю как сломать ваше приложение)))) Достаточно сделать экспорт во время какого-то сложного состояния. А потом сделать импорт и с высокой долей вероятности мы окажемся в ситуации, когда не сможем выйти из этого состояния. Ну или получим некорректное поведение с некорректными ответами.

Krayseer replied:
Теперь состояние не выгружается а хранится в памяти в сервисе UserStateService

vpyzhyanov commented on src/main/java/ru/anykeyers/services/FilterService.java:
```diff
@@ -0,0 +1,18 @@
+import ru.anykeyers.domain.User;
+
+/**
+ * Интерфейс сервиса для фильтрации контактов
```
1. не надо воды, это не диплом, все и так видят, что это интерфейс по ключевому слову `interface`
2. Описание не соответствует имени интерфейса. Название уж очень общее. Что-то в стиле `public interface Service`, т.е. совершенно невозможно догадаться что он фильтрует
3. Судя по описанию метод должен называться `FilterContactService`

Krayseer replied:
исправили

vpyzhyanov commented on src/main/java/ru/anykeyers/services/FilterService.java:
```diff
@@ -0,0 +1,18 @@
+     * Отфильтровать контакты в зависимости от состояния пользователя и введенного значения
+     * @param user пользователь
+     * @param value значение, по которому будет происходить фильтрация
+     * @return результат фильтрации контактов
```
атрибут `@return` нужен для того, чтобы указать конкретные возвращаемые значения, если это возможно, или описание в каком виде возвращаются значения. Вы же использовали атрибут не по назначению, написали что-то в стиле "спасибо КЭП". А на самом деле вообще непонятно что же это за `результат фильтрации контактов` и почему он в виде строки.

P.S. Если не можете написать полезную информацию, ничего не пишите, даже сам `@return`.

Krayseer replied:
хорошо, в дальнейшем так и будем делать

vpyzhyanov commented on src/main/java/ru/anykeyers/services/SearchService.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Интерфейс сервиса для поиска контактов и групп
+ */
+public interface SearchService {
```
аналогично

vpyzhyanov replied:
Не исправлено! Да и в этом сервисе нет ни одного метода, который связан с группами, зато в описании про них написано.

vpyzhyanov commented on src/main/java/ru/anykeyers/services/SearchService.java:
```diff
@@ -0,0 +1,36 @@
+    /**
+     * Найти все контакты пользователя
+     * @param user пользователь
+     * @return список контактов
```
странно, что они в виде строки. Я бы ожидал коллекцию контактов. Подозреваю что вы путаете фильтрацию с форматированнием. Отсюда подозреваю вас в переходе на тёмную сторону ООП.

Krayseer replied:
Исправили сервисы, теперь они возвращают какие то данные (объекты, коллекции) или void

vpyzhyanov commented on src/main/java/ru/anykeyers/services/SortService.java:
```diff
@@ -0,0 +1,18 @@
+/**
+ * Интерфейс сервиса по сортировке контактов
+ */
+public interface SortService {
```
все замечания аналогичные

vpyzhyanov commented on src/main/java/ru/anykeyers/services/SortService.java:
```diff
@@ -0,0 +1,18 @@
+     * @param value аргумент сортировки
+     * @return результат сортировки контактов
+     */
+    String sortByUserStateAndKind(User user, String value);
```
здесь тоже смешали сортировку с форматированнием в строку

vpyzhyanov commented on src/main/java/ru/anykeyers/services/impl/FilterServiceImpl.java:
```diff
@@ -0,0 +1,160 @@
+
+    public FilterServiceImpl(ContactRepository contactRepository) {
+        this.contactRepository = contactRepository;
+        this.messages = new Messages();
```
очередное считывание одного и того же файла

vpyzhyanov commented on src/main/java/ru/anykeyers/services/impl/FilterServiceImpl.java:
```diff
@@ -0,0 +1,160 @@
+    private Predicate<Contact> getGenderPredicate(String value) {
+        switch (value) {
+            case FilterGenderKind.MAN -> {
+                return contact -> contact.getGender() != null && contact.getGender().equals(Gender.MAN);
```
`&& ... ` лучше перенести на следующую строку

vpyzhyanov commented on src/main/java/ru/anykeyers/services/impl/FilterServiceImpl.java:
```diff
@@ -0,0 +1,160 @@
+                return contact -> contact.getGender() != null && contact.getGender().equals(Gender.MAN);
+            }
+            case FilterGenderKind.WOMAN -> {
+                return contact -> contact.getGender() != null && contact.getGender().equals(Gender.WOMAN);
```
`&& ...`  лучше перенести на следующую строку

vpyzhyanov commented on src/main/java/ru/anykeyers/services/impl/SearchServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+                ? messages.getMessageByKey("commons.empty")
+                : contacts.stream()
+                    .map(Contact::getInfo)
+                    .collect(Collectors.joining("\n\n"));
```
совершенно точно этот класс перегружен функционалом

Krayseer replied:
вынесено в UiContactMapper

vpyzhyanov commented on src/main/java/ru/anykeyers/services/impl/SearchServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+     * Получить информацию контакта по имени пользователя и имени контакта
+     * @param user пользотватель
+     * @param name имя контакта
+     * @return результат поиска
```
спасибо КЭП)))

vpyzhyanov commented on src/main/java/ru/anykeyers/services/impl/SearchServiceImpl.java:
```diff
@@ -0,0 +1,117 @@
+                        return new Object[]{ contact.getName(), phoneNumber };
+                    })
+                    .map(contactInfo -> messages.getMessageByKey("contact.info", contactInfo))
+                    .collect(Collectors.joining("\n\n"));
```
совершенно точно этот класс перегружен функционалом

vpyzhyanov commented on src/main/java/ru/anykeyers/services/impl/SortServiceImpl.java:
```diff
@@ -0,0 +1,56 @@
+        return contacts.stream()
+                .sorted(isReversed ? comparator.reversed() : comparator)
+                .map(Contact::getInfo)
+                .collect(Collectors.joining("\n\n"));
```
совершенно точно этот класс перегружен функционалом

vpyzhyanov reviewed (changes requested):
1. Не исправлены замечания из предыдущей задачи
2. **Нет тестов на новые функции.** Ребята, как так???
3. Перешли на тёмную сторону ООП

Снял баллы за git из-за этого:

![image](https://github.com/Krayseer/contacter/assets/71599019/284dbc5f-7f9d-4e56-9762-91158700c242)

Похоже за это надо снять баллы за гит (выделено красным). Тут можно предъявить неиспользование веток (раз ветка не попала в итоговый код), работу только одного человека (раз только ваша ветка была смёрджена в task2)
То, что выделено жёлтым выглядит очень подозрительно. У двух разных людей абсолютно одинаковые комментарии к коммиту. Либо это комментарии не соответствующие действительности, либо это попытка симулировать работу двух человек над одной задачей, когда фактически её делал только один человек.

Krayseer committed:
Большие изменения по архитектуре
1) Сервисы теперь не возвращают строковые результаты, а объекты или void
2) Добавлены собственные ошибки и их обработчики для выдачи сообщения пользователю
3) У пользователя убрано поле StateInfo, которое хранило информацию о состоянии. Теперь для этого создан класс UserStateService, который хранит в себе состояния пользователей
4) Добавлены и обновлены тесты
5) Переработана документация

Avrooid committed:
Добавлены тесты для обработчика состояния выполнения операций

vpyzhyanov commented on src/main/java/ru/anykeyers/processor/state/impl/GroupStateProcessor.java:
```diff
@@ -0,0 +1,108 @@
+        });
+        registerHandler(State.EDIT_GROUP_FIELD, (user, field) -> {
+            switch (field.toLowerCase()) {
+                case GroupEditActionKind.NAME -> {
```
Уместнее было определить enum, а значение получать так: `GroupEditActionKind.values()[num]`

Krayseer replied:
казалось что пробег по всем элементам будет затратный

vpyzhyanov replied:
Какой пробег? Где здесь пробег по всем элементам? В упор такого не вижу)

vpyzhyanov commented on src/main/java/ru/anykeyers/repository/file/FileContactRepository.java:
```diff
@@ -0,0 +1,83 @@
+
+    @Override
+    public void delete(Contact contact) {
+        Optional.of(findByUsername(contact.getUsername()))
```
неуместное использование Optional. null тут никогда не будет. Похоже забыли убрать

Krayseer replied:
забыли убрать

vpyzhyanov commented on src/main/java/ru/anykeyers/service/SortContactService.java:
```diff
@@ -0,0 +1,23 @@
+/**
+ * Сервис по сортировке контактов
+ */
+public interface SortContactService {
```
Опять тёмная сторона ООП. Размазали ответственность. У вас есть сервис `Сервис для работы с контактами`, но по какой-то причине сортировка **контактов** никак не связана с **контактами**.

Krayseer replied:
Согласен, исправим

vpyzhyanov commented on src/main/java/ru/anykeyers/service/FilterContactService.java:
```diff
@@ -0,0 +1,23 @@
+/**
+ * Сервис по фильтрации контактов
+ */
+public interface FilterContactService {
```
ещё сильнее размазали ответственность

Krayseer replied:
Исправим

vpyzhyanov commented on src/main/java/ru/anykeyers/service/SearchService.java:
```diff
@@ -0,0 +1,22 @@
+/**
+ * Сервис по поиску контактов и групп
+ */
+public interface SearchService {
```
нереально сильно размазали

Krayseer replied:
Исправим

vpyzhyanov commented on src/main/java/ru/anykeyers/service/DataRetrievalService.java:
```diff
@@ -0,0 +1,38 @@
+/**
+ * Сервис по получению полной информации контактов, групп
+ */
+public interface DataRetrievalService {
```
А тут сразу 2 тёмные стороны сразу. И всё та же размазанная ответственность и класс тянет на себе слишком много, работает и с контактами и с группами

Krayseer replied:
Исправим

vpyzhyanov commented on src/main/java/ru/anykeyers/processor/state/impl/OperationStateProcessor.java:
```diff
@@ -0,0 +1,254 @@
+                return ex.getMessage();
+            }
+            userStateInfo.clear();
+            return getUiContactsInfo(contacts);
```
не самое удачное имя метода. Операция 'get' обычно подразумевает возвращение какого-то объекта или свойства. Тут же вы преобразуете объект в строку. Лучше подойдёт `toFormattedString` или 'formatContacts' или просто 'format'

Krayseer replied:
поправим

vpyzhyanov commented on src/main/java/ru/anykeyers/service/impl/UserStateServiceImpl.java:
```diff
@@ -0,0 +1,28 @@
+        if (!stateInfoByUsername.containsKey(user.getUsername())) {
+            stateInfoByUsername.put(user.getUsername(), new StateInfo());
+        }
+        return stateInfoByUsername.get(user.getUsername());
```
это всё можно сделать в одну строчку так: `stateInfoByUsername.computeIfAbsent(user.getUsername(), s -> new StateInfo());`

Krayseer replied:
поправим

vpyzhyanov replied:
> это всё можно сделать в **одну** строчку

Я смотрю вы бездумно принимаете предложения, да и невнимательно читаете комментарий. И даже не видите к какой строчке написано замечание. Ну JavaDoc даже не потрудились почитать.

В общем убирайте этот `stateInfoByUsername.get(user.getUsername())` отсюда. computeIfAbsent уже возвращает элемент.

vpyzhyanov reviewed (changes requested):
По прежнему есть критичные замечания по ООП

Krayseer committed:
Убраны классы по поиску, фильтрации и сортировке, методы из них вынесены в ContactService и GroupService.
Для ContactService были написаны дополнительные классы, доступные только из пакета, где находится ContactServiceImpl
Классы с выбором действий теперь перечисления(enum), для получения действия берётся индекс

vpyzhyanov commented:
Ребята, вы разве хотите лишиться даже уже полученных баллов за gti? Обнаружен прямой коммит в ветку task2

![image](https://github.com/Krayseer/contacter/assets/71599019/e18be11b-58a2-404b-90ba-0218cf4dc83f)

В третьей задаче рекомендую личные ветки тоже отребейзить (может даже повторно смёрджить), чтобы все коммиты и ветки выглядели как положено.

vpyzhyanov commented on src/main/java/ru/anykeyers/common/Utils.java:
```diff
@@ -0,0 +1,40 @@
+     * @param field     строковое представление числа
+     * @return конкретный аргумент перечисления
+     */
+    public Enum<?> getEnumKindByField(Enum<?>[] enumerate, String field) {
```
Почему же вы не используете дженерики там где они нужны?

```java
public <T extends Enum<T>> Enum<T> getEnumKindByField(Enum<T>[] enumerate, String field) {
```

И никаких кастов делать не надо!

Krayseer replied:
это то что нужно! спасибо!

vpyzhyanov commented on src/main/java/ru/anykeyers/common/Utils.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * Утилитарный класс
+ */
+public final class Utils {
```
Не бывает просто утилитарных классов, все утилиты должны быть разделены по группам находится в своих классах. Например `EnumUtils`

Krayseer replied:
Поправим

vpyzhyanov commented on src/main/java/ru/anykeyers/common/Utils.java:
```diff
@@ -0,0 +1,40 @@
+            num = Integer.parseInt(field);
+            return enumerate[num - 1];
+        } catch (NumberFormatException numberFormatException) {
+            throw new InvalidNumberFormat();
```
Нет ни одной причины по которой вы должны создать свой тип исключения. Единственная причина создать своё исключение - чтобы его перехватить. Но у вас этого нет. Вы им всё равно не пользуетесь. Могли просто добавить сообщение ` throw new IllegalArgumentException("???", numberFormatException);` или, если стек сохранять не нужно, то просто ` throw new NumberFormatException("???");`

Krayseer replied:
хорошо, будет поправлено

vpyzhyanov commented on src/main/java/ru/anykeyers/common/Utils.java:
```diff
@@ -0,0 +1,40 @@
+        } catch (NumberFormatException numberFormatException) {
+            throw new InvalidNumberFormat();
+        } catch (IndexOutOfBoundsException indexOutOfBoundsException) {
+            throw new BadArgumentException();
```
аналогично

vpyzhyanov reviewed (changes requested):
Задача принята. Как исправите последние замечания, можете мёрджить.

Krayseer committed:
Правки по замечаниям.
Теперь в EnumUtils используются дженерики => убраны ненужные касты
